File: 卡码/代码随想录/JS/二叉树/06-二叉搜索树修改与构造/LeetCode_0450.js
//450. 删除二叉搜索树中的节点
import TreeNode from "../TreeNode.js"

const deleteNode = function(root, key) {
  if(!root) return null;
  if(key<root.val) root.left=deleteNode(root.left,key);
  else if(key>root.val) root.right=deleteNode(root.right,key);
  else{
    if(!root.left&&!root.right) return null;
    else if(root.left&&!root.right) return root.left;
    else if(!root.left&&root.right) return root.right;
    else{
      let parent=root;
      let tmp=root.right;
      while(tmp.left){
        parent=tmp;
        tmp=tmp.left;
      }
      root.val=tmp.val;
      if(parent==root) parent.right=tmp.right;
      else parent.left=tmp.right;
    }
  }
  return root;
};

File: 卡码/代码随想录/1/06-二叉树/C++/06-二叉搜索树修改与构造/LeetCode_0450.cpp
//450. 删除二叉搜索树中的节点
#include <iostream>
#include "../TreeNode.h"
using namespace std;

class Solution {
public:
  TreeNode* deleteNode(TreeNode* root, int key) {
  	if(!root) return nullptr;
  	if(root->val>key) root->left=deleteNode(root->left,key);
  	else if(root->val<key) root->right=deleteNode(root->right,key);
  	else{
			if(!root->left&&!root->right) return nullptr;
			else if(!root->left&&root->right) return root->right;
			else if(root->left&&!root->right) return root->left;
			else{
				TreeNode* tmp=root->right;
				TreeNode* parent=root;
				while(tmp->left){
					parent=tmp;
					tmp=tmp->left;
				}
				root->val=tmp->val;
				if(parent==root) parent->right=tmp->right;
				else parent->left=tmp->right;
				delete(tmp);
			}
		}
		return root;
  }
};

int main(){
  system("pause");
  return 0;
}
